asda.py

```diff
@@ -0,0 +1,519 @@
+        self.vx = self.vx.T
+        self.vy = self.vy.T
+
+        # reform data is factor is greater than 1
+        if self.factor > 1:
+            self.vx = reform2d(vx, self.factor)
+            self.vy = reform2d(vy, self.factor)
+
+        # Initialise Gamma1 and Gamma2
+        self.gamma = np.array([np.zeros_like(self.vx),
+                               np.zeros_like(self.vy)]).T
+
+        # pm vectors, see equation (8) in Graftieaux et al. 2001 or Equation
+        # (1) in Liu et al. 2019
+        pm = np.array([[i, j]
+                       for i in np.arange(-self.r, self.r + 1)
+                       for j in np.arange(-self.r, self.r + 1)], dtype=float)
+
+        # mode of vector pm
+        pnorm = np.linalg.norm(pm, axis=1)
+
+        # Number of points in the concerned region
+        N = (2 * self.r + 1) ** 2
+
+        # Create index array
+        index=np.array([[i, j]
+                        for i in np.arange(self.r, self.dshape[0]-self.r)
+                        for j in np.arange(self.r, self.dshape[1]-self.r)]).T
+
+        # Generate velocity field
+        vel = gen_vel(index[1], index[0])
+
+        # Iterate over the array gamma
+        for dim, (i, j) in enumerate(product(np.arange(self.r,
+                                                       self.dshape[0] - self.r,
+                                                       1),
+                                             np.arange(self.r,
+                                                       self.dshape[1] - self.r,
+                                                       1))):
+
+            self.gamma[i, j, 0], \
+                self.gamma[i, j, 1] = calc_gamma(pm, vel[..., dim], pnorm, N)
+
+        # Transpose back vx & vy
+        self.vx = vx
+        self.vy = vy
+
+        return self.gamma
+
+    def center_edge(self, rmin=4, gamma_min=0.89):
+        '''
+        Find all swirls from gamma1, and gamma2
+
+        Inputs:
+            rmin - minimum radius of swirls, all swirls with radius less than
+                   rmin will be rejected.
+            gamma_min - minimum value of gamma1, all potential swirls with
+                   peak gamma1 values less than gamma_min will be rejected.
+
+        Outputs:
+            center: center location of vortices, in the form of [x, y]
+            edge: edge location of vortices, in the form of [x, y]
+            points: all points within vortices, in the form of [x, y]
+            peak: maximum/minimum gamma1 value in vortices
+            radius: equivalent radius of vortices
+            All results are in pixel coordinates
+        '''
+
+        # Initial dictionary setup
+        self.edge_prop = {'center': (), 'edge': (), 'points': (), 'peak': (),
+                          'radius': ()}
+
+        # Turn interactive plotting off
+        plt.ioff()
+        plt.figure(-1)
+        # Find countours
+        cs = plt.contour(self.gamma[..., 1], levels=[-2 / np.pi, 2 / np.pi])

```

Comment:
That would be prettier for sure. I don't like the plt.ioff(), plt.figure(-1) lines here. Absolutely unnecessary matplotlib figure initialisation. Will do this.

Reply:
I am doing this now

Reply:
skimage.measure.find-contours does not give the same results on real data (it gives the same result on the artificial data). So will leave this part with using plt.contour

Reply:
I would look into that some more, I wouldn't trust one over the other really.

Reply:
The problem with scikit.measure.find_contours is it also includes open contours which we don't want. There is no such option to omit all open contours. We will need to write our own...

Reply:
I would imagine you could filter out the open ones reasonably easily.

> (The closed-ness of a contours can be tested by checking whether the beginning point is the same as the end point.)

Reply:
a np.where statement would probably be a efficient vecotrized method of doing that.

Reply:
I don't think it worth the effort. Matplotlib contour was used in the paper, so i would just use continue to use that...

Reply:
It's really a hack to use matplotlib like that, there is *so* much overhead in doing it, and it would be easy to make matplotlib show the hidden figure etc.

Reply:
No worries guys, I will try to figure out using scikit...

Reply:
I used it in SSC. I will be there in few minutes and will help you out.